akka-projection-grpc/src/main/scala/akka/projection/grpc/consumer/scaladsl/EventConsumer.scala

```diff
@@ -0,0 +1,196 @@
+    override private[akka] def apply(eventEnvelope: EventEnvelope[Any]): EventEnvelope[Any] = {
+      val newPid = f(eventEnvelope)
+      if (newPid eq eventEnvelope.persistenceId) eventEnvelope
+      else
+        new EventEnvelope[Any](
+          offset = eventEnvelope.offset,
+          persistenceId = newPid,
+          sequenceNr = eventEnvelope.sequenceNr,
+          eventOption = eventEnvelope.eventOption,
+          timestamp = eventEnvelope.timestamp,
+          eventMetadata = eventEnvelope.eventMetadata,
+          entityType = eventEnvelope.entityType,
+          slice = eventEnvelope.slice,
+          filtered = eventEnvelope.filtered,
+          source = eventEnvelope.source,
+          tags = eventEnvelope.tags)
+    }
+  }
+
+  private final case class MapPayload(f: EventEnvelope[Any] => Option[Any]) extends Transformation {
+    override private[akka] def apply(eventEnvelope: EventEnvelope[Any]): EventEnvelope[Any] = {
+      val newMaybePayload = f(eventEnvelope)
+      if (newMaybePayload eq eventEnvelope.eventOption) eventEnvelope
+      else {
+        new EventEnvelope[Any](
+          offset = eventEnvelope.offset,
+          persistenceId = eventEnvelope.persistenceId,
+          sequenceNr = eventEnvelope.sequenceNr,
+          eventOption = newMaybePayload,
+          timestamp = eventEnvelope.timestamp,
+          eventMetadata = eventEnvelope.eventMetadata,
+          entityType = eventEnvelope.entityType,
+          slice = eventEnvelope.slice,
+          filtered = newMaybePayload.isEmpty,
+          source = eventEnvelope.source,
+          tags = eventEnvelope.tags)
+      }
+
+    }
+  }
+
+  private final case class AndThen(first: Transformation, next: Transformation) extends Transformation {
+    override private[akka] def apply(eventEnvelope: EventEnvelope[Any]): EventEnvelope[Any] =
+      next.apply(first.apply(eventEnvelope))
+  }
+
+  /**
+   * Transformation of events from the producer type to the internal representation stored in the journal
+   * and seen by local projections.
+   *
+   * Events can be excluded by mapping them to `None`.
+   */
+  sealed trait Transformation {
+    def mapTags[Event](f: EventEnvelope[Event] => Set[String]): Transformation =
+      MapTags(f.asInstanceOf[EventEnvelope[Any] => Set[String]])
+    def mapPersistenceId[Event](f: EventEnvelope[Event] => String): Transformation =
+      MapPersistenceId(f.asInstanceOf[EventEnvelope[Any] => String])
+    def mapPayload[Event, B <: Event](f: EventEnvelope[Event] => Option[B]): Transformation =
+      MapPayload(f.asInstanceOf[EventEnvelope[Any] => Option[B]])
+
+    def andThen(transformation: Transformation): Transformation =
+      AndThen(this, transformation)
+
+    // FIXME do we need async?
+    private[akka] def apply(eventEnvelope: EventEnvelope[Any]): EventEnvelope[Any]
+  }
+
+  def grpcServiceHandler(eventConsumer: EventConsumerDestination)(
```

Comment:
I think something is missing for handling more than one streamId? EventConsumerDestination is for one particular streamId, right?

The existing EventProducer has
```
def grpcServiceHandler(sources: Set[EventProducerSource])
```

Reply:
The destination one or more stream ids, but all will end up in the same journal and use the same filters etc. Wasn't entirely sure it is needed but I have a FIXME somewhere with the same thought.

Reply:
Maybe one stream id per `EventProducerPushDestination` would make more sense actually. I'll do that and have a map in the service to look up which for a given connection.

Reply:
one stream per destination, multiple destinations added in [d72553f](https://github.com/akka/akka-projection/pull/932/commits/d72553fe6523989437228c1c0b6efd63e84c23f8)

Reply:
that also makes it possible to have different transformations and filters for different streams, which I think was missing before